### PR TITLE
API Keys: add AuthCard and listing page

### LINKED ISSUE #36805: Frontend for API Keys

### PR DESCRIPTION
**Pull request by shaunlebron**
Closes https://github.com/metabase/metabase/issues/36805

Adding the entry point for managing API Keys at <http://localhost:3000/admin/settings/authentication>

## Summary

1. Added `ApiKey` type to “metabase-types/api/admin”
1. Added `ApiKeysApi` entry to “metabase/services”
1. Added `AuthCard` for API Keys
1. Added `ManageApiKeys` page
    * Added CRUD modals for creating, editing, and deleting keys.
    * Added secondary modals for regenerating a key and copying an unmasked key.
1. Edited `GroupListing` page to show number of API keys belonging to each group
1. Edited `GroupDetails` page to show each API key belonging the group

I also created some reusable components:
1. Added `FormGroupWidget` component for selecting a single Group.

### CODE REVIEW

**Comment by dpsutton** [comment]
is there a shared feature branch this should target instead of master?

File: e2e/runner/cypress-runner-run-tests.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ const isFolder = !!folder;
 const isOpenMode = args["--open"];
 
 const getSourceFolder = folder => {
-  return `./e2e/test/scenarios/${folder}/**/*.cy.spec.js`;
+  return `./e2e/test/scenarios/${folder}/**/*.cy.spec.{js,ts}`;
```
Comment (iethree) [inline e2e/runner/cypress-runner-run-tests.js:15]: is there a separate PR in master you can target allowing .ts specs? If not, could you pull this out as a separate thing?

File: e2e/test/scenarios/admin-2/api-keys.cy.spec.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+describe("scenarios > admin > settings > API keys", () => {
+  // TODO: replace intercepts below with actual requests to test backend
+  let mockRows: ApiKey[] = [];
+
+  beforeEach(() => {
+    restore();
+    mockRows = [...MOCK_ROWS];
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    cy.signInAsAdmin();
```
Comment (iethree) [inline e2e/test/scenarios/admin-2/api-keys.cy.spec.ts:85]: 🤔 another reason for a separate PR to enable typescript, I have a pretty strong aversion to using // @ts-ignore.

 I think we need to define types for our plugins and tasks before enabling typescript specs
 
 https://docs.cypress.io/guides/tooling/typescript-support#Types-for-plugins

File: e2e/runner/cypress-runner-run-tests.js (hunk 1 above)
Comment (shaunlebron) [reply e2e/runner/cypress-runner-run-tests.js]: oops thanks, I had already done that and forgot to rebase

Comment (shaunlebron) [reply e2e/runner/cypress-runner-run-tests.js]: Here’s the PR that was merged: https://github.com/metabase/metabase/pull/36813

File: e2e/test/scenarios/admin-2/api-keys.cy.spec.ts (hunk 2 above)
Comment (shaunlebron) [reply e2e/test/scenarios/admin-2/api-keys.cy.spec.ts:85]: Added here, thanks: https://github.com/metabase/metabase/pull/36958

File: frontend/src/metabase/forms/components/FormGroupWidget/FormGroupWidget.tsx
Comment (iethree) [inline frontend/src/metabase/forms/components/FormGroupWidget/FormGroupWidget.tsx:1]: 1. is there any chance this component is useful outside of the admin interface? if not, I'd put it under admin?
2. is there anywhere else we're selecting groups that we could use this component? could this work in the create user modal? This feels like it could be it's own PR

File: frontend/src/metabase/components/CopyTextInput/CopyTextInput.styled.tsx
Comment (iethree) [inline frontend/src/metabase/components/CopyTextInput/CopyTextInput.styled.tsx:1]: this could maybe also be it's own little PR, maybe implementing it in another existing form or two?

File: frontend/src/metabase/plugins/builtin.js
Comment (iethree) [inline frontend/src/metabase/plugins/builtin.js:1]: I see you've followed the existing pattern here, but using "plugins" for non-enterprise features feels overly complicated to me. Maybe see if there's any other opinions on this in #enginerring-frontend ?

File: frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            {!isShowingEmptyTable && (
+              <Text color="text.1">{t`Allow users to use the API keys to authenticate their API calls.`}</Text>
+            )}
+          </Stack>
+          <Button
+            variant="filled"
+            onClick={() => setModal("create")}
+          >{t`Create API Key`}</Button>
+        </Group>
+        <LoadingAndErrorWrapper loading={isLoading} error={error}>
```
Comment (iethree) [inline frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx:84]: nice use of this wrapper 👍

File: frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                          setActiveApiKey(apiKey);
+                          setModal("delete");
+                        }}
+                      />
+                    </Group>
+                  </td>
+                </tr>
+              ))}
+            </tbody>
+          </table>
```
Comment (iethree) [inline frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx:133]: can this table be its own view component? would probably make it easier to unit test

File: frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+          {isShowingEmptyTable && (
+            <Stack
+              h="40rem" // TODO: how to make this fill only available window height?
+              align="center"
+              justify="center"
+              spacing="sm"
+            >
+              <Title>{t`No API keys here yet`}</Title>
+              <Text color="text.1">{t`Create API keys to programmatically authenticate their API calls.`}</Text>
+            </Stack>
```
Comment (iethree) [inline frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx:143]: this could also be a nice little <ApiKeysEmptyState /> component

File: frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.unit.spec.tsx
Comment (iethree) [inline frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.unit.spec.tsx:1]: 💖

**Review by iethree — COMMENTED** [review]
This looks really great!  🎈 

Left some preliminary comments on how to split it up a bit and break down some of the bigger components.

File: e2e/test/scenarios/admin-2/api-keys.cy.spec.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+      "fetchKeys",
+    );
+    cy.intercept("POST", "/api/api-key", req => {
+      mockRows.push(req.body);
+      req.reply(200);
+    });
+
+    cy.visit("/admin/settings/authentication/api-keys");
+    cy.wait("@fetchKeys");
+    cy.button("Create API Key").click();
```
[the remark this answers is not in the document]
Comment (shaunlebron) [reply e2e/test/scenarios/admin-2/api-keys.cy.spec.ts]: db55f0116d3ff971096c751b40ee656d782aec23

File: frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx (hunk 4 above)
Comment (shaunlebron) [reply frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx]: 09418becd34e8c2e6e332286a9ac8646b5b77ab8

Comment (shaunlebron) [reply frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx]: 1ab2574247898627b92b896316cbfffaf2cb42d8

File: frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx (hunk 5 above)
Comment (shaunlebron) [reply frontend/src/metabase/admin/settings/components/ApiKeys/ManageApiKeys.tsx]: 599f8a3ed7a2c37af85ab4bb7b0dae78274a9ddc

File: frontend/src/metabase/components/CopyTextInput/CopyTextInput.styled.tsx
Comment (shaunlebron) [reply frontend/src/metabase/components/CopyTextInput/CopyTextInput.styled.tsx:1]: https://github.com/metabase/metabase/pull/37021

**Comment by shaunlebron** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/36258?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #37021** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37021?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #36258** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36258?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/36258?utm_source=stack-comment).

**Comment by shaunlebron** [comment]
I split this PR up into smaller chunks as requested:

* `master`
  * **PR #37095** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37095?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #37096** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37096?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #37097** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37097?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #37098** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37098?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #37099** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37099?utm_source=stack-comment-icon" target="_blank">Graphite</a>
